source/common/upstream/subset_lb.cc

```diff
@@ -0,0 +1,497 @@
+#include "common/upstream/subset_lb.h"
+
+#include <unordered_set>
+
+#include "envoy/runtime/runtime.h"
+
+#include "common/common/assert.h"
+#include "common/config/metadata.h"
+#include "common/config/well_known_names.h"
+#include "common/protobuf/utility.h"
+#include "common/upstream/load_balancer_impl.h"
+#include "common/upstream/ring_hash_lb.h"
+
+#include "api/cds.pb.h"
+
+namespace Envoy {
+namespace Upstream {
+
+SubsetLoadBalancer::SubsetLoadBalancer(LoadBalancerType lb_type, HostSet& host_set,
+                                       const HostSet* local_host_set, ClusterStats& stats,
+                                       Runtime::Loader& runtime, Runtime::RandomGenerator& random,
+                                       const LoadBalancerSubsetInfo& subsets)
+    : lb_type_(lb_type), stats_(stats), runtime_(runtime), random_(random),
+      fallback_policy_(subsets.fallbackPolicy()), default_subset_(subsets.defaultSubset()),
+      subset_keys_(subsets.subsetKeys()), original_host_set_(host_set),
+      original_local_host_set_(local_host_set) {
+  ASSERT(subsets.isEnabled());
+
+  // Create filtered default subset (if necessary) and other subsets based on current hosts.
+  update(host_set.hosts(), {}, false);
+
+  // Configure future updates.
+  host_set.addMemberUpdateCb([this](const std::vector<HostSharedPtr>& hosts_added,
+                                    const std::vector<HostSharedPtr>& hosts_removed) {
+    update(hosts_added, hosts_removed, false);
+  });
+
+  if (local_host_set) {
+    // No need to invoke update with local = true: subset creation handles the local hosts.
+    local_host_set->addMemberUpdateCb([this](const std::vector<HostSharedPtr>& hosts_added,
```

Comment:
I believe you will need to hold a handle here and remove if this cluster/LB gets deleted, otherwise you will crash when the local host set updates. See load balancer base.